pkg/server/settingsworker_test.go

```diff
@@ -375,3 +379,112 @@ func TestSettingsZipkinBatchSize(t *testing.T) {
+	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{})
+	defer tc.Stopper().Stop(ctx)
+
+	execStatement := func(statement string) {
+		if _, err := tc.ServerConn(rand.Intn(3)).Exec(statement); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	execStatement(`CREATE DATABASE t`)
+	execStatement(`CREATE TABLE test (k INT PRIMARY KEY, v INT)`)
+
+	// Set up tracing collector and variable to count the number of traces
+	// collected from INSERT INTO statements
+	nTraces := int64(0)
+	collectorHandlerFcn := func(rw http.ResponseWriter, req *http.Request) {
+		bodyBytes, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return
+		}
+		t := thrift.NewTMemoryBuffer()
+		if _, err = t.Write(bodyBytes); err != nil {
+			return
+		}
+		p := thrift.NewTBinaryProtocolTransport(t)
+		_, batchSize, err := p.ReadListBegin()
+		if err != nil {
+			return
+		}
+
+		for i := 0; i < batchSize; i += 1 {
+			span := zipkincore.Span{}
+			if err = span.Read(p); err != nil {
+				return
+			}
+
+			if span.Name == `exec stmt` {
+				for _, annotation := range span.Annotations {
+					if strings.Contains(annotation.Value, `INSERT INTO test VALUES`) &&
+						strings.Contains(annotation.Value, `in state: Open`) {
+						atomic.AddInt64(&nTraces, 1)
+					}
+				}
+			}
+		}
+	}
+	traceCollectorServer := httptest.NewServer(http.HandlerFunc(collectorHandlerFcn))
+	defer traceCollectorServer.Close()
+
+	// Configure tracing in the test cluster
+	traceCollectorServerURL, err := url.Parse(traceCollectorServer.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+	setClusterTraceSamplingRate := func(rate float64) {
+		execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.sampling.rate = %f`, rate))
+	}
+	execStatement(`SET CLUSTER SETTING trace.zipkin.collector.batch.size = 100`)
+	execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.collector = '%s'`, traceCollectorServerURL.Host))
+
+	// Function that writes 1000 rows
+	startKey := 0
+	write10000Rows := func() {
+		waitGroup := sync.WaitGroup{}
+		for key := startKey; key < startKey+1000; key += 10 {
+			waitGroup.Add(1)
+			go func(wg *sync.WaitGroup, k int) {
+				defer wg.Done()
+				for v := 0; v < 10; v++ {
+					execStatement(fmt.Sprintf(`INSERT INTO test VALUES (%d, %d)`, k+v, v))
+				}
+			}(&waitGroup, key)
+		}
+		waitGroup.Wait()
+		startKey = startKey + 1000
+	}
+
+	// Test 100% Sampling Rate
+	setClusterTraceSamplingRate(1.0)
+	atomic.StoreInt64(&nTraces, 0)
+	write10000Rows()
+	testutils.SucceedsSoon(t, func() error {
+		currentTraces := atomic.LoadInt64(&nTraces)
+		if currentTraces != 1000 {
+			return fmt.Errorf(`could not trace all SQL 'INSERT INTO' statements. traced %d out of 10000`, currentTraces)
+		}
+		return nil
+	})
+
+	// Test 50% Sampling Rate
+	setClusterTraceSamplingRate(0.5)
+	atomic.StoreInt64(&nTraces, 0)
+	write10000Rows()
+	testutils.SucceedsSoon(t, func() error {
+		currentTraces := atomic.LoadInt64(&nTraces)
+		if currentTraces < 450 && currentTraces > 550 {
```

Comment:
This has a bit of a race condition which may result in a false positive. It is possible that `currentTraces` may exceed 550 after we pass this test due to `nTraces` being incremented asynchronously. In testing, this basically never happened, which makes me feel safer about this.